### PR TITLE
EVG-17487: SpruceForm for SpawnHostModal

### PR DESCRIPTION
**Pull request by SupaJoon**
[EVG-17487
](https://jira.mongodb.org/browse/EVG-17487)
### Description
test it with this link: 
http://localhost:3000/spawn/host?spawnHost=True&distroId=ubuntu1804-workstation&taskId=evergreen_ubuntu1604_dist_patch_33016573166a36bd5f46b4111151899d5c4e95b1_5ecedafb562343215a7ff297_20_05_27_21_39_46
### Screenshots
Screen Shot 2022-09-21 at 3 35 23 PM

### Evergreen PR
https://github.com/evergreen-ci/evergreen/pull/5954

### CODE REVIEW

File: src/components/SpruceForm/Widgets/DistroDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
+import styled from "@emotion/styled";
```
Comment (khelif96) [inline src/components/SpruceForm/Widgets/DistroDropdown.tsx:1]: This feels like to specific of a component to be a SpruceForm component. SpruceForm is meant to be generic and reusable to create forms anywhere on spruce we shouldn't have a widget that is specific to one form. I think this should just be a generic SearchableDropdown widget.

File: src/pages/projectSettings/tabs/PluginsTab/PluginsTab.tsx [hunk 2]
```diff
@@ -30,7 +30,7 @@ export const PluginsTab: React.VFC<TabProps> = ({
       getFormSchema(
         projectType === ProjectType.AttachedProject ? repoData : null
       ),
-    [projectType, repoData, formData]
+    [projectType, repoData]
```
Comment (khelif96) [inline src/pages/projectSettings/tabs/PluginsTab/PluginsTab.tsx:33]: Why did this change it doesn't seem related to the rest of the changes in here.

File: src/pages/spawn/spawnHost/spawnHostModal/transformer.test.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      noExpiration: false,
+      volumeId: null,
+      homeVolumeSize: 504,
+      publicKey: {
+        name: "a name woo",
+        key: "blah blahsart",
+      },
+      savePublicKey: true,
+      distroId: "ubuntu1804-workstation",
+      region: "rofl-east",
```
Comment (khelif96) [inline src/pages/spawn/spawnHost/spawnHostModal/transformer.test.ts:59]: I agree with annies comment on the other pr we should probably use a real region to make future projects easier.

File: src/pages/spawn/spawnHost/spawnHostModal/transformer.test.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      })
+    ).toStrictEqual(data.mutationInput);
+  });
+  const data = {
+    formData: {
+      distro: {
+        value: "ubuntu1804-workstation",
+        isVirtualWorkstation: true,
+      },
+      region: "rofl-east",
```
Comment (khelif96) [inline src/pages/spawn/spawnHost/spawnHostModal/transformer.test.ts:19]: same comment

**Review by khelif96 — CHANGES_REQUESTED** [review]
Great job so far had a few initial comments

File: src/pages/spawn/spawnHost/spawnHostModal/getFormSchema.tsx [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+
+interface LoadTaskDataOntoHostLabelProps {
+  options: {
+    buildVariant: string;
+    taskDisplayName: string;
+    revision: string;
+  };
+}
+
+export const LeafyGreenCheckboWithCustomLabel: React.VFC<
```
Comment (sophstad) [inline src/pages/spawn/spawnHost/spawnHostModal/getFormSchema.tsx:33]: I'd move this into its own file. It may also make more sense as a field.

File: src/pages/spawn/spawnHost/spawnHostModal/transformer.test.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        expiration: "Thu Dec 08 2022 14:52:51 GMT-0500 (Eastern Standard Time)",
+      },
+      homeVolumeDetails: {
+        selectExistingVolume: false,
+        volumeSize: 504,
+        volumeSelect: "",
+      },
+    },
+    mutationInput: {
+      isVirtualWorkStation: true,
```
Comment (sophstad) [inline src/pages/spawn/spawnHost/spawnHostModal/transformer.test.ts:47]: We could probably test some more inputs here like not using a virtual workstation, no expiration, etc.

**Review by sophstad — CHANGES_REQUESTED** [review]

File: src/components/SpruceForm/Widgets/LeafyGreenWidgets.tsx [hunk 7]
```diff
@@ -95,25 +97,26 @@ const WarningText = styled.p`
...
-}) => {
-  const { "data-cy": dataCy, marginBottom, tooltipDescription } = options;
+export const LeafyGreenCheckBox: React.VFC<
+  SpruceWidgetProps & { customLabel?: JSX.Element }
+> = ({ value, label, onChange, disabled, options, readonly, customLabel }) => {
+  const {
+    "data-cy": dataCy,
+    marginBottom,
+    tooltipDescription,
+    elementWrapperCSS,
```
[the remark this answers is not in the document]
Comment (sophstad) [reply src/components/SpruceForm/Widgets/LeafyGreenWidgets.tsx]: Yeah I like this new approach! 🎉

[the remark this answers is not in the document]
Comment (khelif96) [reply src/components/SpruceForm/Widgets/LeafyGreenWidgets.tsx]: Great idea SGTM!

File: src/components/Spawn/utils.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+}: GetNoExpirationCheckboxTooltipCopyProps) =>
+  disableExpirationCheckbox
+    ? `You have reached the max number of unexpirable ${
+        isVolume ? `volumes` : `hosts`
+      }  (${limit}). Toggle an existing ${
+        isVolume ? "volume" : "host"
+      } to expirable to enable this checkbox.`
+    : undefined;
+
+export const validateTask = (taskData: GetSpawnTaskQuery["task"]) => {
```
Comment (khelif96) [inline src/components/Spawn/utils.ts:21]: nit. Should this live in the spawnHost/utils since i don't think this is used by volumes

File: src/components/SpruceForm/Widgets/types.ts [hunk 9]
```diff
@@ -12,6 +14,8 @@ export interface SpruceWidgetProps extends WidgetProps {
     showLabel: boolean;
     tooltipDescription: string;
     warnings: string[];
+    elementWrapperCSS: SerializedStyles;
+    type: TextInputType;
```
[the remark this answers is not in the document]
Comment (khelif96) [reply src/components/SpruceForm/Widgets/types.ts:16]: Should we make this optional?

File: src/components/SpruceForm/Widgets/LeafyGreenWidgets.tsx [hunk 10]
```diff
@@ -95,25 +97,26 @@ const WarningText = styled.p`
...
       <Checkbox
         data-cy={dataCy}
         checked={value}
         label={
           <>
-            {label}
+            {customLabel || label}
             {tooltipDescription && (
               <Tooltip
+                popoverZIndex={zIndex.max_do_not_use_tooltip}
```
[the remark this answers is not in the document]
Comment (khelif96) [reply src/components/SpruceForm/Widgets/LeafyGreenWidgets.tsx:119]: The `zIndex`'s are established in a way where you would be able to have a tooltip on top of a modal by default.

`zIndex.modal = 10` and `zIndex.tooltip = 20` so a tooltip should always  be able to appear on top of the modal. 

`zIndex.max_do_not_use` is a special modal that we use for onboarding and should take precedence over any other content on the page.  We shouldn't use it for normal modals.

File: src/components/Spawn/ExpirationField.tsx [hunk 11]
```diff
@@ -103,7 +103,7 @@ export const ExpirationField: React.VFC<ExpirationFieldProps> = ({
           >
             <span>
               <PaddedCheckbox
-                data-cy="neverExpireCheckbox"
+                data-cy="never-expire-checkbox"
                 disabled={disableExpirationCheckbox}
                 label="Never"
                 checked={noExpiration}
```
Comment (khelif96) [inline src/components/Spawn/ExpirationField.tsx:109]: This doesn't check automatically when ever i try to spawn a virtual workstation like it currently does.

File: src/pages/spawn/spawnHost/spawnHostModal/Widgets/DistroDropdown.tsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+  padding: 0;
+`;
+const OptionContainer = styled.div`
+  padding: ${size.xs};
+`;
+const Option = styled(OptionContainer)`
+  word-break: break-word;
+  cursor: pointer;
+  :hover {
+    background-color: ${uiColors.gray.light1};
```
[the remark this answers is not in the document]
Comment (khelif96) [reply src/pages/spawn/spawnHost/spawnHostModal/Widgets/DistroDropdown.tsx:121]: It's more of a consistency thing since we destructure ui colors every where else in the codebase.

File: package.json [hunk 13]
```diff
@@ -108,6 +108,7 @@
     "lodash.every": "4.6.0",
     "lodash.get": "4.4.2",
     "lodash.isequal": "4.5.0",
+    "lodash.omit": "4.5.0",
```
[the remark this answers is not in the document]
Comment (khelif96) [reply package.json:111]: One thing i noticed is that `lodash.omit` is not properly typed so we lose type safety if we want to use this in the future.

image

 Would you be open to replacing it with a typed alternative and adding it to our utils?
I came up with this
```ts
const omit = <T extends object, K extends [...(keyof T)[]]>(
  obj: T,
  params: K
) => {
  const newObj = { ...obj };
  params.forEach((param) => delete newObj[param]);
  return newObj as Omit<T, K[number]>;
};
```

image

File: src/gql/queries/spawn-task.graphql [hunk 14]
```diff
@@ -1,7 +1,7 @@
 #import "../fragments/baseTask.graphql"
 
-query GetSpawnTask($taskId: String!, $execution: Int) {
-  task(taskId: $taskId, execution: $execution) {
+query GetSpawnTask($taskId: String!) {
+  task(taskId: $taskId, execution: 0) {
```
Comment (khelif96) [inline src/gql/queries/spawn-task.graphql:4]: Why did we make this change? i think the old behavior was getting the latest task execution while this always gets the first execution.

File: src/pages/spawn/spawnHost/SpawnHostButton.tsx [hunk 15]
```diff
@@ -8,23 +8,24 @@ import { PlusButton } from "components/Buttons";
...
-import { SpawnHostModal } from "pages/spawn/spawnHost/index";
+import { useSpruceConfig, useUpdateURLQueryParams } from "hooks";
+import { SpawnHostModal } from "pages/spawn/spawnHost/SpawnHostModal";
 import { HostStatus } from "types/host";
 import { queryString } from "utils";
 import { QueryParams } from "../types";
 
 const { parseQueryString } = queryString;
-export const SpawnHostButton = () => {
+export const SpawnHostButton: React.VFC = () => {
```
Comment (khelif96) [inline src/pages/spawn/spawnHost/SpawnHostButton.tsx:18]: nit. can we just add a space here.

File: src/components/Spawn/ExpirationField.tsx (hunk 11 above)
Comment (khelif96) [reply src/components/Spawn/ExpirationField.tsx:106]: Thanks!

**Review by khelif96 — CHANGES_REQUESTED** [review]
I'm still getting the bug with setting an expiration date i can't seem to set any date.

https://user-images.githubusercontent.com/4605522/194633281-deb0010d-2887-403c-8e3b-2f07b65ec10a.mov

It would also be awesome if we could replace those other dropdowns with Leafygreen dropdowns  😄

File: src/pages/spawn/spawnHost/SpawnHostButton.tsx [hunk 16]
```diff
@@ -8,23 +8,24 @@ import { PlusButton } from "components/Buttons";
...
 import { GET_MY_HOSTS } from "gql/queries";
-import { useSpruceConfig } from "hooks";
-import { SpawnHostModal } from "pages/spawn/spawnHost/index";
+import { useSpruceConfig, useUpdateURLQueryParams } from "hooks";
+import { SpawnHostModal } from "pages/spawn/spawnHost/SpawnHostModal";
 import { HostStatus } from "types/host";
 import { queryString } from "utils";
 import { QueryParams } from "../types";
 
 const { parseQueryString } = queryString;
```
Comment (khelif96) [inline src/pages/spawn/spawnHost/SpawnHostButton.tsx:17]: It was merely to add a space between this declaration and the component declaration below.
```suggestion
const { parseQueryString } = queryString;

```

File: src/pages/spawn/spawnHost/SpawnHostButton.tsx (hunk 15 above)
Comment (khelif96) [reply src/pages/spawn/spawnHost/SpawnHostButton.tsx:18]: Done

File: src/components/SpruceForm/Widgets/DateTimePicker.tsx [hunk 17]
```diff
@@ -20,10 +20,12 @@ export const DateTimePicker: React.VFC<
     ? utcToZonedTime(new Date(value), timezone)
     : new Date(value);
   const isDisabled = disabled || readonly;
-  const handleChange = (d: Date) =>
+  const handleChange = (d: Date) => {
+    console.log("on h");
```
Comment (khelif96) [inline src/components/SpruceForm/Widgets/DateTimePicker.tsx:24]: console.log

File: src/utils/object/omit.test.ts [hunk 18]
```diff
@@ -0,0 +1,14 @@
+import { omit } from "utils/object";
+
+describe("omit", () => {
+  it("returns a new object without the supplied key", () => {
+    expect(omit<{ b?: number }, ["b"]>({}, ["b"])).toMatchObject({});
```
Comment (khelif96) [inline src/utils/object/omit.test.ts:5]: can you remove the typing here? it makes reading it confusing. 

Can you add another test with multiple keys and show it removing only a few of them.

File: src/utils/object/omit.test.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+    expect(omit({ a: 1 }, ["a"])).toMatchObject({});
+  });
+  it("returns a new object reference even when the input and outupt object have deep equality", () => {
```
Comment (khelif96) [inline src/utils/object/omit.test.ts:8]:
```suggestion
  it("returns a new object reference even when the input and output object have deep equality", () => {
```

Comment (khelif96) [reply src/utils/object/omit.test.ts:8]: This test is a bit confusing is it trying to test that the input and output are the same? 
It doesn't even actually test out the functionality since its not actually removing any fields from inputObj.

File: src/components/SpruceForm/Widgets/DateTimePicker.tsx [hunk 20]
```diff
@@ -41,6 +43,8 @@ export const DateTimePicker: React.VFC<
       {description && <Description>{description}</Description>}
       <DateTimeContainer>
         <DatePicker
+          // @ts-ignore
```
Comment (khelif96) [inline src/components/SpruceForm/Widgets/DateTimePicker.tsx:46]: nit. 
```suggestion
          // @ts-expect-error
```

File: src/components/SpruceForm/Widgets/DateTimePicker.tsx [hunk 21]
```diff
@@ -49,6 +53,8 @@ export const DateTimePicker: React.VFC<
           disabledDate={disabledDate}
         />
         <TimePicker
+          // @ts-ignore
```
Comment (khelif96) [inline src/components/SpruceForm/Widgets/DateTimePicker.tsx:56]: nit 
```suggestion
          // @ts-expect-error
```

File: src/components/SpruceForm/Widgets/DateTimePicker.tsx [hunk 22]
```diff
@@ -20,10 +20,12 @@ export const DateTimePicker: React.VFC<
     ? utcToZonedTime(new Date(value), timezone)
```
Comment (khelif96) [inline src/components/SpruceForm/Widgets/DateTimePicker.tsx:20]: value is initially null which throws this warning

image

image
Its probably easiest to set a default value for it until we load the data

**Review by khelif96 — COMMENTED** [review]
image
I am starting to get this warning do you know what its from?

File: src/pages/spawn/spawnHost/SpawnHostButton.tsx [hunk 23]
```diff
@@ -8,23 +8,25 @@ import { PlusButton } from "components/Buttons";
...
     GET_MY_HOSTS
   );
 
   const spruceConfig = useSpruceConfig();
   const { search } = useLocation();
-  const queryParams = parseQueryString(search);
-  const shouldSpawnHost = queryParams[QueryParams.SpawnHost] === "True";
-  const [openModal, setOpenModal] = useState(shouldSpawnHost);
+  const initialOpen =
+    parseQueryString(search)[QueryParams.SpawnHost] === "True";
```
Comment (khelif96) [inline src/pages/spawn/spawnHost/SpawnHostButton.tsx:28]: can't wait to bring in `useQueryParam()` from parsley

File: src/pages/spawn/spawnHost/spawnHostModal/Widgets/DistroDropdown.tsx [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+    "data-cy": dataCy,
+    ariaLabelledBy,
+    elementWrapperCSS,
+    enumOptions,
+  } = options;
+
+  const searchableOptions = enumOptions.reduce<OptionValue[]>(
+    (accum, { schema, value }) => {
+      const { isVirtualWorkstation } = schema;
+      accum[isVirtualWorkstation ? 0 : 1].distros.push({
```
Comment (khelif96) [inline src/pages/spawn/spawnHost/spawnHostModal/Widgets/DistroDropdown.tsx:50]: I understand what this is doing now but do you think we could make it more verbose or add some comments so its more clear in the future when we come back to this.

File: src/utils/object/omit.test.ts [hunk 25]
```diff
@@ -0,0 +1,16 @@
+import { omit } from "utils/object";
+
+describe("omit", () => {
+  it("returns an object without the supplied key(s)", () => {
+    expect(omit({} as { a?: number }, ["a"])).toMatchObject({});
```
Comment (khelif96) [inline src/utils/object/omit.test.ts:5]: I think we could omit this test. This type assertion is misleading since it doesn't affect the real type of the object.

File: src/utils/object/omit.test.ts [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+    expect(omit({ a: 1 }, ["a"])).toMatchObject({});
+    expect(omit({ a: 1, b: 1 }, ["b"])).toMatchObject({ a: 1 });
+    expect(omit({ a: 1, b: 1, c: 1 }, ["b", "c"])).toMatchObject({ a: 1 });
+  });
+  it("returns a new object reference", () => {
+    const inputObj = {};
+    const resultObj = omit<{ b?: number }, ["b"]>(inputObj, ["b"]);
+    expect(resultObj).toMatchObject(inputObj);
+    expect(resultObj).not.toBe(inputObj);
```
Comment (khelif96) [inline src/utils/object/omit.test.ts:14]: Same comment

**Review by khelif96 — APPROVED** [review]
Approved mod the pending comments! 🎉

File: src/utils/object/omit.test.ts (hunk 26 above)
Comment (khelif96) [reply src/utils/object/omit.test.ts:14]: Sorry wrong line i was referring to line 12
